product_variant_default_code/models/product.py

```
@@ -15,30 +15,91 @@
+
+
+def extract_token(s):
+    pattern = re.compile(r'\[([^\]]+?)\]')
+    return set(pattern.findall(s))
+
+
+def sanitize_reference_mask(product, mask):
+    tokens = extract_token(mask)
+    attribute_names = set()
+    for line in product.attribute_line_ids:
+        attribute_names.add(line.attribute_id.name)
+    if not tokens.issubset(attribute_names):
+        raise except_orm(_('Error'), _('Found unrecognized attribute name in '
+                                       '"Variant Reference Mask"'))
+
+
+def render_default_code(product, mask):
+    product_attrs = defaultdict(str)
+    reference_mask = ReferenceMask(mask)
+    for value in product.attribute_value_ids:
+        if value.attribute_code:
+            product_attrs[value.attribute_id.name] += value.attribute_code
+    all_attrs = extract_token(mask)
+    missing_attrs = all_attrs - set(product_attrs.keys())
+    missing = dict.fromkeys(missing_attrs, PLACE_HOLDER_4_MISSING_VALUE)
+    product_attrs.update(missing)
+    default_code = reference_mask.safe_substitute(product_attrs)
+    product.default_code = default_code
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    default_prefix = fields.Char(
-        string='Reference prefix',
-        help='Prefix for using when building internal references for the '
-             'variants generated from this template')
+    reference_mask = fields.Char(
+        string='Variant reference mask',
+        help='Reference mask for building internal references of a'
+             'variant generated from this template.'
+             'eg. "PREFIX-[r]-[d]-[f]-APPENDIX" where the "[]" surrounded '
+             'string like "r,d,f" are attribute name for the current product '
+             'When rendering, the "[xxx]" part will be replaced '
+             'by the corresponding code of attribute value for the variant.\n'
+             'Note: make sure characters "[,]" do not appear in your '
+             'attribute name')
```

@pedrobaeza 

> We should provide a default value for this mask

I don't get it. 

The default value for the mask will be automatically created as long as the product is created with attributes.
let say you create a product named 'A' with two attributes, 'Size' and 'Color' , such as:

Product:  `A`
`Color`: Red(r), Yellow(y), Black(b)             #Red, Yellow, Black  are the attribute value, 'r', 'y', 'b' are the corresponding code
`Size`: L(l), XL(x)

The automatically generated default value for the mask will be `[Color]-[Size]`

and then the  'default code' on the variants will be something like `r-l` `b-l` `r-x` ...

If you like you can change the mask value whatever you like, you can even have the attribute name appear more than once in the mask such as ,  'fancyA/[Size]~[Color]~[Size]', when saved the default code on variants will be something like `fancyA/l~r~l`(for variant with Color "Red" and Size "L") `fancyA/x~y~x` (for variant with Color "Yellow" and Size "XL")

I hope it will be a bit clearer now, I will update the example in help.